aiohttp/web_urldispatcher.py

```diff
@@ -174,6 +179,63 @@ def url(self, *, filename, query=None):
         url = self._prefix + filename
         return self._append_query(url, query)
 
+    def _sendfile_cb(self, fut, out_fd, in_fd, offset, count, loop):
+        try:
+            n = os.sendfile(out_fd, in_fd, offset, count)
+        except (BlockingIOError, InterruptedError):
+            return
+        except Exception as exc:
+            loop.remove_writer(out_fd)
+            fut.set_exception(exc)
+            return
+        else:
+            loop.remove_writer(out_fd)
+
+        if n < count:
+            loop.add_writer(out_fd, self._sendfile_cb, fut, out_fd, in_fd,
+                            offset + n, count - n, loop)
+        else:
+            fut.set_result(None)
+
+    @asyncio.coroutine
+    def _sendfile(self, out_fd, in_fd, offset, count, loop):
+        fut = asyncio.Future(loop=loop)
+        loop.add_writer(out_fd, self._sendfile_cb, fut, out_fd, in_fd, offset,
+                        count, loop)
+        return fut
+
+    @asyncio.coroutine
+    def sendfile_system(self, resp, fobj):
+        """
+        Write the content of `fobj` to `resp` using the ``sendfile`` system
+        call.
+
+        `fobj` should be an open file object.
+
+        `resp` should be a :obj:`aiohttp.web.StreamResponse` instance.
+        """
+        yield from resp.drain()
+        req = resp._req
+        loop = req.app.loop
+        out_fd = req._transport._sock_fd
+        in_fd = fobj.fileno()
+        yield from self._sendfile(out_fd, in_fd, 0, os.fstat(in_fd).st_size,
+                                  loop)
+
+    @asyncio.coroutine
+    def sendfile_fallback(self, resp, fobj):
+        """
+        Mimic the :meth:`sendfile` method, but without using the ``sendfile``
+        system call. This should be used on systems that don't support
+        ``sendfile``.
+        """
+        yield from resp.drain()
```

Comment:
To flush any headers, and give the client a chance to bail out early when the chunk_size is large.